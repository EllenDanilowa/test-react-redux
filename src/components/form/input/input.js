import React from 'react';
import PropTypes from 'prop-types';
import {
  InputElement,
  ErrorMessage
} from './input.styled';
import {FieldWrapper, Label} from '../form.styled';

const DEFAULT_ERROR_MESSAGE = 'Oops are you sure? – this doesn’t look right';

const Input = (props) => {
  const {
    name,
    placeholder,
    title,
    type,
    error,
    refFunc,
    errorMessage
  } = props;

  return (
    <FieldWrapper>
      <Label htmlFor={name}>{title}</Label>
      <InputElement
        id={name}
        name={name}
        type={type}
        placeholder={placeholder}
        ref={refFunc}
        error={Boolean(error)}
      />
      {error && <ErrorMessage>{errorMessage || DEFAULT_ERROR_MESSAGE}</ErrorMessage>}
    </FieldWrapper>
  );
};

Input.propTypes = {
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  title: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  refFunc: PropTypes.func.isRequired,
  error: PropTypes.object,
  errorMessage: PropTypes.string
};

Input.defaultProps = {
  type: 'text'
};

export default Input;
