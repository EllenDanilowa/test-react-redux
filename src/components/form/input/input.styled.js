import styled from 'styled-components';
import PropTypes from 'prop-types';
import colors from '../../../styles/colors';

const inputColor = '#495057';
const inputBorderColor = '#ced4da';
const inputBorderFocusedColor = '#80bdff';
const inputShadowFocusedColor = 'rgba(0,123,255,.25)';
const inputErrorColor = '#e21219';
const inputBackgroundErrorColor = '#fdf3f3';

export const InputElement = styled.input`
  background-color: ${(props) => props.error ? inputBackgroundErrorColor : colors.main_color};
  border: 1px solid ${(props) => props.error ? inputErrorColor : inputBorderColor};
  border-radius: 4px;
  box-sizing: border-box;
  color: ${inputColor};
  line-height: 1.5;
  outline: 0;
  padding: 12px 6px;
  width: 100%;
  
  &:focus {
    border-color: ${inputBorderFocusedColor};
    outline: 0;
    box-shadow: 0 0 3px 3px ${inputShadowFocusedColor};
  }
`;
InputElement.propTypes = {
  error: PropTypes.bool,
};

InputElement.defaultProps = {
  error: false
};

export const ErrorMessage = styled.p`
  color: ${inputErrorColor};
  font-size: 12px;
  margin: 4px 0;
`;
