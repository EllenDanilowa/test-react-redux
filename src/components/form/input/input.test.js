import React from 'react';
import {shallow} from 'enzyme';
import Input, {DEFAULT_ERROR_MESSAGE} from './input';
import {InputElement, ErrorMessage} from './input.styled';
import {Label} from '../form.styled';

describe('Input', () => {
  let element;
  let title;
  let name;
  let ref;
  let placeholder;
  let type;
  let error;
  let errorMessage;

  const createInput = ({ref, ...props}) => {
    return shallow(<Input refFunc={ref} {...props}/>)
  }

  beforeEach(() => {
    title = 'Title';
    name = 'name';
    ref = () => {};
    placeholder = 'placeholder';
    type = 'number';
    error = {};
    errorMessage = 'error message';

    element = createInput({ref, name, title, placeholder, type, error, errorMessage});
  });

  it('sets htmlFor prop and title to label', () => {
    const label = element.find(Label);

    expect(label.prop('htmlFor')).toBe(name);
    expect(label.text()).toBe(title);
  });

  it('sets props to input', () => {
    const input = element.find(InputElement);

    expect(input.prop('id')).toBe(name);
    expect(input.prop('name')).toBe(name);
    expect(input.getElement().ref).toBe(ref);
    expect(input.prop('placeholder')).toBe(placeholder);
    expect(input.prop('type')).toBe(type);
    expect(input.prop('error')).toBe(true);
  });

  it('shows error message if error', () => {
    expect(element.find(ErrorMessage).text()).toBe(errorMessage);
  });

  it('shows default error message if error', () => {
    element = createInput({ref, name, title, error});

    expect(element.find(ErrorMessage).text()).toBe(DEFAULT_ERROR_MESSAGE);
  });

  it('sets empty placeholder, type text and hide errors if not exist', () => {
    element = createInput({ref, name, title});
    const input = element.find(InputElement);

    expect(input.prop('placeholder')).toBe('');
    expect(input.prop('type')).toBe('text');
    expect(input.prop('error')).toBe(false);
    expect(element.find(ErrorMessage).length).toBe(0);
  });
});
