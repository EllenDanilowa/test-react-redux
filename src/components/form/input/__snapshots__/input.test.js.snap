// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders component correctly 1`] = `
<div
  className="sc-AxhCb gssqqQ"
>
  <label
    className="sc-AxiKw gfimZx"
    htmlFor="name"
  >
    Title
  </label>
  <input
    className="sc-AxjAm cPjdGB"
    id="name"
    name="name"
    placeholder="placeholder"
    type="number"
  />
  <p
    className="sc-AxirZ liZJXo"
  >
    error message
  </p>
</div>
`;

exports[`Input renders component with default props correctly 1`] = `
<div
  className="sc-AxhCb gssqqQ"
>
  <label
    className="sc-AxiKw gfimZx"
    htmlFor="name"
  >
    Title
  </label>
  <input
    className="sc-AxjAm iCpIIE"
    id="name"
    name="name"
    placeholder=""
    type="text"
  />
</div>
`;

exports[`Input renders component with no error message if no error 1`] = `
<div
  className="sc-AxhCb gssqqQ"
>
  <label
    className="sc-AxiKw gfimZx"
    htmlFor="name"
  >
    Title
  </label>
  <input
    className="sc-AxjAm iCpIIE"
    id="name"
    name="name"
    placeholder="placeholder"
    type="number"
  />
</div>
`;
